zarr/indexing.py

```diff
@@ -0,0 +1,676 @@
+                start = 0
+            else:
+                start = self.chunk_nitems_cumsum[dim_chunk_ix - 1]
+            stop = self.chunk_nitems_cumsum[dim_chunk_ix]
+            dim_out_sel = slice(start, stop)
+
+            yield ChunkDimProjection(dim_chunk_ix, dim_chunk_sel, dim_out_sel)
+
+
+class IntArrayDimIndexer(object):
+    """Integer array selection against a single dimension."""
+
+    def __init__(self, dim_sel, dim_len, dim_chunk_len):
+
+        # ensure array
+        dim_sel = np.asanyarray(dim_sel)
+
+        # check number of dimensions
+        if dim_sel.ndim != 1:
+            raise IndexError('selection must be a 1d array')
+
+        # check dtype
+        if dim_sel.dtype.kind not in 'ui':
+            raise IndexError('selection must be an integer array')
+
+        # handle wraparound
+        loc_neg = dim_sel < 0
+        if np.any(loc_neg):
+            dim_sel[loc_neg] = dim_sel[loc_neg] + dim_len
+
+        # handle out of bounds
+        if np.any(dim_sel < 0) or np.any(dim_sel >= dim_len):
+            raise IndexError('selection contains index out of bounds')
+
+        # handle non-monotonic indices
+        dim_sel_chunk = dim_sel // dim_chunk_len
+        if np.any(np.diff(dim_sel) < 0):
+            self.is_monotonic = False
+            # sort indices to group by chunk
+            self.dim_sort = np.argsort(dim_sel_chunk)
+            self.dim_sel = np.take(dim_sel, self.dim_sort)
+
+        else:
+            self.is_monotonic = True
+            self.dim_sort = None
+            self.dim_sel = dim_sel
+
+        # store attributes
+        self.dim_len = dim_len
+        self.dim_chunk_len = dim_chunk_len
+        self.nchunks = int(np.ceil(self.dim_len / self.dim_chunk_len))
+        self.nitems = len(self.dim_sel)
+
+        # precompute number of selected items for each chunk
+        # note: for dense integer selections, the division operation here is the bottleneck
+        self.chunk_nitems = np.bincount(dim_sel_chunk, minlength=self.nchunks)
+        self.chunk_nitems_cumsum = np.cumsum(self.chunk_nitems)
+        self.dim_chunk_ixs = np.nonzero(self.chunk_nitems)[0]
+
+    def __iter__(self):
+
+        for dim_chunk_ix in self.dim_chunk_ixs:
+
+            # find region in output
+            if dim_chunk_ix == 0:
+                start = 0
+            else:
+                start = self.chunk_nitems_cumsum[dim_chunk_ix - 1]
+            stop = self.chunk_nitems_cumsum[dim_chunk_ix]
+            if self.is_monotonic:
+                dim_out_sel = slice(start, stop)
+            else:
+                dim_out_sel = self.dim_sort[start:stop]
+
+            # find region in chunk
+            dim_offset = dim_chunk_ix * self.dim_chunk_len
+            dim_chunk_sel = self.dim_sel[start:stop] - dim_offset
+
+            yield ChunkDimProjection(dim_chunk_ix, dim_chunk_sel, dim_out_sel)
+
+
+def slice_to_range(s):
+    return range(s.start, s.stop, 1 if s.step is None else s.step)
```

Comment:
Use `slice.indices()` instead to get start/stop/step (this is especially important for tricky cases like negative steps). You'll also need the size of the array dimension.

Reply:
Ah, didn't know about that, nice.